File: tools/server_observer/account.py
from conflict_interface.data_types.hub_types.hub_game import HubGameProperties
from conflict_interface.interface.hub_interface import HubInterface
from conflict_interface.interface.online_interface import OnlineInterface
from tools.server_observer.proxy import Proxy

MAXIMUM_NUMBER_OF_GAMES = 10

def ensure_games_loaded(func):
    def wrapper(self, *args, **kwargs):
        if self.games is None:
            self.get_my_games()
        return func(self, *args, **kwargs)

    return wrapper

class Account:
    username: str
    password: str
    email: str
    proxy_id: str
    proxy_url: str
    hub_itf: HubInterface = None
    games: list[HubGameProperties] = None

    def __repr__(self):
        return f"Account(username={self.username}, proxy_id={self.proxy_id})"

    def to_dict(self):
        return {
            "username": self.username,
            "password": self.password,
            "email": self.email,
            "proxy_id": self.proxy_id,
            "proxy_url": self.proxy_url
        }

    @classmethod
    def from_dict(cls, obj):
        return cls(
            username=obj["username"],
            password=obj["password"],
            email=obj["email"],
            proxy_id=obj["proxy_id"],
            proxy_url=obj["proxy_url"]
        )

    def __init__(self, username, password, email, proxy_id, proxy_url):
        self.username = username
        self.password = password
        self.email = email
        self.proxy_id = proxy_id
        self.proxy_url = proxy_url
        self.hub_itf = HubInterface({
            "http": proxy_url,
            "https": proxy_url,
        })

    def login(self) -> bool:
        if not self.hub_itf.auth:
            return self.hub_itf.login(self.username, self.password)
        return True

    def get_interface(self) -> HubInterface:
        if not self.hub_itf.auth:
            self.hub_itf.login(self.username, self.password)
        return self.hub_itf

    def reset_interface(self):
        self.hub_itf = HubInterface({
            "http": self.proxy_url,
            "https": self.proxy_url,
        })
        self.games = None
        self.hub_itf.login(self.username, self.password)


    def set_proxy(self, proxy: Proxy):
        self.proxy_id = proxy.id
        self.proxy_url = proxy.proxy_url

        self.hub_itf.set_proxy({
            "http": self.proxy_url,
            "https": self.proxy_url,
        })

    @ensure_games_loaded
    def has_maximum_games(self) -> bool:
        """
        Returns True if the account has the maximum number of games assigned to it.
        Hence, it cannot join new games until some old games are finished.
    
        :return: If the account has the maximum number of games assigned to it.
        """
        return len(self.games) >= MAXIMUM_NUMBER_OF_GAMES

    @ensure_games_loaded
    def open_game_slots(self) -> int:
        return MAXIMUM_NUMBER_OF_GAMES - len(self.games)

    def get_my_games(self) -> list[HubGameProperties]:
        if self.games is None:
            if not self.hub_itf.auth:
                self.login()
            self.games = self.hub_itf.get_my_games()
        return self.games

    def has_game(self, game_id: int) -> bool:
        return any([game_id == my_game.game_id for my_game in self.get_my_games()])


    @ensure_games_loaded
    def join_game(self, game_id: int, guest: bool = False, replay_filename: str = None) -> OnlineInterface:
        game_itf = self.hub_itf.join_game(game_id, guest, replay_filename)
        self.games = self.hub_itf.get_my_games()
        return game_itf

File: tools/server_observer/src/account.cpp
#include "account.hpp"
#include <iostream>

Account::Account(const std::string& username, const std::string& password,
                const std::string& email,
                std::shared_ptr<ProxyConfig> proxy_config)
    : username(username)
    , password(password)
    , email(email)
    , proxy_config(proxy_config)
    , hub_interface_(nullptr)
    , games_loaded_(false)
{}

Account::~Account() {
}

bool Account::login_internal() {
    if (!hub_interface_) {
        std::string proxy_url = proxy_config ? proxy_config->to_url() : "";
        hub_interface_ = std::make_shared<HubInterfaceWrapper>(proxy_url, proxy_url);
    }
    
    if (hub_interface_->is_authenticated()) {
        return true;
    }
    
    return hub_interface_->login(username, password);
}

bool Account::login() {
    std::lock_guard<std::mutex> lock(mutex_);
    return login_internal();
}

std::shared_ptr<HubInterfaceWrapper> Account::get_interface() {
    std::lock_guard<std::mutex> lock(mutex_);

    if (!hub_interface_) {
        std::string proxy_url = proxy_config ? proxy_config->to_url() : "";
        hub_interface_ = std::make_shared<HubInterfaceWrapper>(proxy_url, proxy_url);
    }
    
    if (!hub_interface_->is_authenticated()) {
        login_internal();
    }
    
    return hub_interface_;
}

void Account::reset_interface() {
    std::lock_guard<std::mutex> lock(mutex_);
    std::string proxy_url = proxy_config ? proxy_config->to_url() : "";
    hub_interface_ = std::make_shared<HubInterfaceWrapper>(proxy_url, proxy_url);
    games_.clear();
    games_loaded_ = false;
    login_internal();
}

void Account::reset_proxy(std::shared_ptr<ProxyConfig> new_proxy_config) {
    std::lock_guard<std::mutex> lock(mutex_);
    proxy_config = new_proxy_config;
    std::string proxy_url = proxy_config ? proxy_config->to_url() : "";
    hub_interface_ = std::make_shared<HubInterfaceWrapper>(proxy_url, proxy_url);
    games_.clear();
    games_loaded_ = false;
    login_internal();
}

std::vector<HubGameProperties> Account::get_my_games() {
    std::lock_guard<std::mutex> lock(mutex_);
    if (!games_loaded_) {
        ensure_games_loaded_internal();
    }
    return games_;
}

bool Account::has_game(int game_id) {
    std::lock_guard<std::mutex> lock(mutex_);
    ensure_games_loaded_internal();
    for (const auto& game : games_) {
        if (game.game_id == game_id) {
            return true;
        }
    }
    return false;
}

bool Account::has_maximum_games() {
    std::lock_guard<std::mutex> lock(mutex_);
    ensure_games_loaded_internal();
    return games_.size() >= MAXIMUM_NUMBER_OF_GAMES;
}

int Account::open_game_slots() {
    std::lock_guard<std::mutex> lock(mutex_);
    ensure_games_loaded_internal();
    return MAXIMUM_NUMBER_OF_GAMES - games_.size();
}

void Account::ensure_games_loaded_internal() {
    if (games_loaded_) {
        return;
    }
    
    // Ensure hub_interface_ is initialized
    if (!hub_interface_) {
        get_interface();
    }
    
    if (!hub_interface_->is_authenticated()) {
        login_internal();
    }
    
    games_ = hub_interface_->get_my_games();
    games_loaded_ = true;
}

void Account::ensure_games_loaded() {
    std::lock_guard<std::mutex> lock(mutex_);
    ensure_games_loaded_internal();
}

json Account::to_json() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return {
        {"username", username},
        {"password", password},
        {"email", email},
        {"proxy_id", proxy_config ? proxy_config->proxy_id : ""},
        {"proxy_url", proxy_config ? proxy_config->to_url() : ""}
    };
}

std::shared_ptr<Account> Account::from_json(const json& j) {
    std::string proxy_url = j.value("proxy_url", "");
    auto proxy_config = std::make_shared<ProxyConfig>(ProxyConfig::from_url(proxy_url));
    proxy_config->proxy_id = j.value("proxy_id", "");

    return std::make_shared<Account>(
        j.value("username", ""),
        j.value("password", ""),
        j.value("email", ""),
        proxy_config
    );
}
